### ISSUE TITLE
Prevent parallel execution of prompt functions

### ISSUE THREAD

**Issue by zzamboni** [elves/elvish#601]
When `$edit:-prompts-max-wait` is set, it seems that a new copy of the prompt function is started with every keystroke. If the prompt function is slow or heavy, this can have undesirable effects. For example, with git-status indicators in the prompt, typing in a large git repo can result in the following errors while typing:
```
prompt function error: failed to create pipe: pipe: too many open files
prompt function error: failed to create pipe: pipe: too many open files
...
```
Looking at the processes reveals a large number of git processes running at the same time. If what I'm trying to run is a git command, it often happens that the command fails because of the status commands running on the repository (i.e. "lock file exists" errors).

I think it would be good if only one prompt function was allowed to be running at a time.

**Comment by xiaq** [comment]
Good point. The implementation will require some interesting synchronization because we also want to gurantee freshness of the prompt, so we do not want refresh requests to simply be dropped when refreshing is in progress.

**Comment by zzamboni** [comment]
I would propose to keep it as simple as possible. In my opinion, lazy background update is good enough for a prompt display. I.e. start the prompt function upon a keystroke, only if there's no other one running. As soon as it finishes, update the prompt.

I.e. if the prompt function takes 1 second, and I am typing at 5 keystrokes per second, the function would only be invoked every 5 keystrokes.
